### PR TITLE
Add full git commit hash to release channel manifests

### PR DESCRIPTION
**Pull request by SimonSapin**
The full hash is necessary to build the download URL for "alternate" compiler builds. This is a first step for https://github.com/rust-lang-nursery/rustup.rs/issues/1099.

### CODE REVIEW

File: src/bootstrap/dist.rs [hunk 1]
```diff
@@ -364,7 +364,9 @@ impl Step for Rustc {
         cp("README.md");
         // tiny morsel of metadata is used by rust-packaging
         let version = build.rust_version();
+        let sha = build.rust_sha().unwrap_or("");
         t!(t!(File::create(overlay.join("version"))).write_all(version.as_bytes()));
+        t!(t!(File::create(overlay.join("git-commit-hash"))).write_all(sha.as_bytes()));
```
Comment (alexcrichton) [inline src/bootstrap/dist.rs]: This `version` file is currently present in a number of tarballs that we produce, and from a consistency point of view it seems odd for the hash to only be in one? Would it be possible to refactor this to ensure there's a git commit file in all the tarballs?

File: src/tools/build-manifest/src/main.rs [hunk 2]
```diff
@@ -425,16 +439,15 @@ impl Builder {
...
 
-    fn write_manifest(&self, manifest: &str, name: &str) {
-        let dst = self.output.join(name);
-        t!(t!(File::create(&dst)).write_all(manifest.as_bytes()));
-        self.hash(&dst);
-        self.sign(&dst);
+    fn write_channel_files(&self, channel_name: &str, manifest: &Manifest) {
+        self.write(&toml::to_string(&manifest).unwrap(), channel_name, ".toml");
+        self.write(&manifest.date, channel_name, "-date.txt");
+        self.write(&manifest.git_commit_hash, channel_name, "-git-commit-hash.txt");
```
Comment (alexcrichton) [inline src/tools/build-manifest/src/main.rs]: Do we for sure need a dedicated file for the git commit? This'll already be present in the manifest, and the `date.txt` was only added much later as a request (ideally this extra file wouldn't exist)

File: src/tools/build-manifest/src/main.rs [hunk 3]
```diff
@@ -107,6 +107,7 @@ static MINGW: &'static [&'static str] = &[
 struct Manifest {
     manifest_version: String,
     date: String,
+    git_commit_hash: String,
```
Comment (alexcrichton) [inline src/tools/build-manifest/src/main.rs]: I think that this commit may be best inside each `Package`? I'd imagine that maybe eventually we'd have a use for the commit information other than just for `rustc` but also for `cargo`

File: src/tools/build-manifest/src/main.rs [hunk 4]
```diff
@@ -382,14 +379,31 @@ impl Builder {
...
+        let output = t!(cmd.output());
+        if !output.status.success() {
             panic!("failed to learn version:\n\n{:?}\n\n{}\n\n{}",
                    cmd,
-                   String::from_utf8_lossy(&version.stdout),
-                   String::from_utf8_lossy(&version.stderr));
+                   String::from_utf8_lossy(&output.stdout),
+                   String::from_utf8_lossy(&output.stderr));
+        }
+        String::from_utf8_lossy(&output.stdout).trim().to_string()
```
Comment (Mark-Simulacrum) [inline src/tools/build-manifest/src/main.rs]: Maybe I'm not quite following, but why the changes to `version()`? It still should be doing the same thing, right?

**Review by Mark-Simulacrum — COMMENTED** [review]
This seems fine in general but there are a few implementation nits. Others can comment more on the rustup/artifact side.

File: src/tools/build-manifest/src/main.rs (hunk 2 above)
Comment (SimonSapin) [reply src/tools/build-manifest/src/main.rs]: We don’t really need it, just like we don’t really need `date.txt`. A separate file makes it marginally easier for Servo’s bootstrap to extract the commit hash, but we can also parse the TOML file.

I can remove it if you prefer.

File: src/bootstrap/dist.rs [hunk 5]
```diff
@@ -364,7 +364,9 @@ impl Step for Rustc {
         cp("README.md");
         // tiny morsel of metadata is used by rust-packaging
         let version = build.rust_version();
+        let sha = build.rust_sha().unwrap_or("");
```
[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/bootstrap/dist.rs]: I made this omit the file at first, but then I didn’t know how to separate “this tarball does not contain this file” from other kinds of errors in `Command::new("tar").arg("xf") …`

Any suggestion?

File: src/tools/build-manifest/src/main.rs (hunk 4 above)
Comment (SimonSapin) [reply src/tools/build-manifest/src/main.rs]: I changed a variable name before copy/pasting the method, but it’s otherwise unchanged. I can revert that if you prefer.

File: src/tools/build-manifest/src/main.rs (hunk 3 above)
Comment (SimonSapin) [reply src/tools/build-manifest/src/main.rs]: Everything under https://s3.amazonaws.com/rust-lang-ci/rustc-builds are in "subdirectories" named after the rust-lang/rust commit hash, even for other packages like Cargo. But if you prefer I could work on that change in case the Cargo commit hash is needed later for some reason.

Comment (SimonSapin) [reply src/tools/build-manifest/src/main.rs]: Figuring out the respective build systems in respective repositories to add text files in various tarballs to make this info available in the first place is non-trivial, though.

File: src/bootstrap/dist.rs (hunk 1 above)
Comment (SimonSapin) [reply src/bootstrap/dist.rs]: Add for source and extended tarballs. I don’t know if the rust-lang/rust commit hash makes sense in Cargo and RLS tarballs, and the rust-lang/cargo and rust-lang-nursery/rls commit hashes are not easily available.

File: src/bootstrap/dist.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/bootstrap/dist.rs]: Maybe it *should* be an error to try to build a manifest with `ignore-git = true`?

File: src/tools/build-manifest/src/main.rs (hunk 3 above)
Comment (SimonSapin) [reply src/tools/build-manifest/src/main.rs]: I’ve made it optional (always missing at the moment for Cargo and RLS) and per-package.

File: src/bootstrap/dist.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (SimonSapin) [reply src/bootstrap/dist.rs]: Switched to not including the file when the info is missing.

File: src/tools/build-manifest/src/main.rs [hunk 6]
```diff
@@ -167,6 +168,9 @@ struct Builder {
     rust_version: String,
     cargo_version: String,
     rls_version: String,
+    rust_git_commit_hash: Option<String>,
+    cargo_git_commit_hash: Option<String>,
```
Comment (alexcrichton) [inline src/tools/build-manifest/src/main.rs]: It looks like these are never initialized?

**Comment by alexcrichton** [comment]
Looks good! I think though that cargo/rls tarballs don't have files with the git commit in them?

File: src/tools/build-manifest/src/main.rs (hunk 6 above)
Comment (SimonSapin) [reply src/tools/build-manifest/src/main.rs]: Fixed.

**Comment by SimonSapin** [comment]
> I think though that cargo/rls tarballs don't have files with the git commit in them?

That’s correct. These would need to be added separately. (If someone finds a use for them, they are not used for https://github.com/rust-lang-nursery/rustup.rs/issues/1099 which motivates this PR.)
